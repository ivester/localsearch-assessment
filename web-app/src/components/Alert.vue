<template>
  <v-alert
    :text="message"
    type="error"
  >
    <!-- div wrapper to show the button under the text and not inline with the text -->
    <div>
      <v-btn class="alert-button" to="/">Go to Search</v-btn>
    </div>
  </v-alert>
</template>

<script lang="ts" setup>
  interface Props {
    message: string;
  }

  defineProps<Props>()
</script>

<style lang="scss" scoped>
  .alert-button {
    margin-top: 16px;
  }
</style>
