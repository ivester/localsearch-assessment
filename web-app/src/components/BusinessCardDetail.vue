<template>
  <v-row>
    <v-col
      cols="12"
      md="6"
    >
      <v-list>
        <v-list-item
          :title="business.where"
          prepend-icon="mdi-map-marker"
          subtitle="Address"
        />
        <v-list-item
          v-if="business?.url"
          :title="business?.urlFormatted || business.url"
          :href="business.url"
          target="_blank"
          prepend-icon="mdi-web"
          subtitle="Website"
        />
        <v-list-item
          v-if="business?.phone"
          :title="business?.phoneFormatted || business.phone"
          :href="`tel:${business.phone}`"
          prepend-icon="mdi-phone"
          subtitle="Phone"
        />
      </v-list>
    </v-col>

    <v-col
      cols="12"
      md="6"
    >
      <BusinessCardOpeningHours
        v-if="business?.openingHours"
        :opening-hours="business.openingHours"
      />
    </v-col>

</v-row>
</template>

<script lang="ts" setup>
  import { Business } from '@/types';
  import BusinessCardOpeningHours from './BusinessCardOpeningHours.vue';

  interface Props {
    business: Business;
  }

  defineProps<Props>()
</script>
