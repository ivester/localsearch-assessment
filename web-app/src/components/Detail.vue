<template>
  <v-container class="fill-height">
    <v-responsive class="align-center text-center fill-height">
      <v-card>
          <v-card-title>
            <span class="headline">{{ business?.name }}</span>
          </v-card-title>
          Address: {{ business?.where }} <br />
          // TODO website
          Website: TODO <br />
          // TODO phone
          Phone: TODO <br />
          Opening Hours: <br />
          <br />

          <!-- TODO add everywhere classes in the BEM style -->
          <span
            v-for="openingHour in openingHoursFormatted"
            :key="openingHour.id"
            class="opening-hours"
          >
            {{openingHour.days}}:
            <!-- Ignoring the key lint error here as I am not worried about performance here. Otherwise I would have to add a unique id to each list element -->
            <!-- https://vuejs.org/guide/essentials/list.html#maintaining-state-with-key -->
            <!-- eslint-disable-next-line vue/require-v-for-key -->
            <span v-for="hour in openingHour.hours">
              {{ hour }}<br>
            </span>
            <br>
          </span>
        </v-card>
    </v-responsive>
  </v-container>
</template>

<script lang="ts" setup>
  import {
    BusinessDetail,
    OpeningHour,
    OpeningHourMerged,
    OpeningHourFormatted,
    Day,
    Hour
  } from '@/types';
  import axios from 'axios'
  import { computed } from 'vue';
  import { ref, onMounted } from 'vue'
  import { useRoute } from 'vue-router'

  const route = useRoute()

  const business = ref<BusinessDetail | null>(null)

  onMounted(async () => {
    // I could load all the data when searching and then just pass the data to the detail view
    // but then I could not reload a detail page on its own
    const { data } = await axios.get<BusinessDetail>(`http://localhost:4000/business/${route.params.id}`)
    business.value = data
  })

  // --- Computed ---

  // days with same hours grouped together
  const openingHoursMerged = computed((): OpeningHourMerged[] =>
    business.value?.openingHours.reduce(mergeOpeningHours, []) || [])

  // days and hours with same hours grouped together and formatted for display
  const openingHoursFormatted = computed((): OpeningHourFormatted[] =>
    // TODO maybe there is a better name for openingHourMerged - I call the param group - maybe openingHourGroupe?
    openingHoursMerged.value.map((group: OpeningHourMerged): OpeningHourFormatted => ({
        ...group,
        days: formatDays(group.days),
        hours: formatHours(group.hours)
      })
    ))

  // --- Methods ---

  // group days with same hours together
  const mergeOpeningHours = (acc: OpeningHourMerged[], curr: OpeningHour): OpeningHourMerged[] => {
    const previous = acc.at(-1)
    // only merge days with same hours together if they follow each other
    const hoursSameAsPervious = JSON.stringify(previous?.hours) === JSON.stringify(curr.hours)

    if (hoursSameAsPervious) {
      // add the current day to previous group if hours are the same
      previous?.days.push(curr.day)
    } else {
      // create new group if hours are not the same
      acc.push({
        id: curr.id,
        days: [curr.day],
        hours: curr.hours
      })
    }

    return acc
  }

  // format days to display the first and last day if multiple days are selected
  const formatDays = (days: Day[]): string => {
    const firstDay = days.at(0)
    const lastDay = days.at(-1)
    const hasMultipleDays = days.length > 1
    const daysFormatted = hasMultipleDays
      ? `${firstDay} - ${lastDay}`
      : firstDay

    return daysFormatted || ''
  }

  // format hours to show the opening hours if defined or the string 'closed' if not defined
  const formatHours = (hours: Hour[]): string[] => hours.length
    ? hours.map(time => `${time.start} - ${time.end}`)
    : ['closed']
</script>

<style lang="scss" scoped>
  .opening-hours {
    text-transform: capitalize;
  }
</style>
