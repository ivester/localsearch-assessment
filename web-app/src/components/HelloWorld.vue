<template>
  <v-container class="fill-height">
    <v-responsive class="align-center text-center fill-height">
      <code>{{ data.displayed_what }}</code>
    </v-responsive>
  </v-container>
</template>

<script lang="ts" setup>
  import axios from 'axios'
  import { onMounted, ref } from 'vue'

  interface Data {
    displayed_what: string
  }

  const data = ref<Data>({ displayed_what: '' })
  onMounted(async () => {
    const response = await axios.get<Data>('http://localhost:4000')
    data.value = response.data
  })
</script>
