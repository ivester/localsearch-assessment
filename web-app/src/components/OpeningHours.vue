<template>
  <div class="opening-hours">
    Opening Hours: <br />
    <br />

    <span
      v-for="openingHour in openingHoursFormatted"
      :key="openingHour.id"
      class="opening-hours__values"
    >
      <span>{{openingHour.days}}: </span>
      <!-- Ignoring the key lint error here as I am not worried about performance here. Otherwise I would have to add a unique id to each list element -->
      <!-- https://vuejs.org/guide/essentials/list.html#maintaining-state-with-key -->
      <!-- eslint-disable-next-line vue/require-v-for-key -->
      <span v-for="hour in openingHour.hours">
        {{ hour }}<br>
      </span>
    </span>
  </div>
</template>

<script lang="ts" setup>
  import {
    OpeningHour,
    OpeningHourGrouped,
    OpeningHourFormatted,
    Day,
    Hour,
  } from '@/types';
  import { computed } from 'vue';

  // --- Props ---

  interface Props {
    openingHours: OpeningHour[];
  }

  const props = defineProps<Props>()

  // --- Computed ---

  // days with same opening hours grouped together
  const openingHoursGrouped = computed((): OpeningHourGrouped[] =>
    props.openingHours.reduce(groupOpeningHours, []) || [])

  // opening hours formatted for display
  const openingHoursFormatted = computed((): OpeningHourFormatted[] =>
    openingHoursGrouped.value.map((group: OpeningHourGrouped): OpeningHourFormatted => ({
        ...group,
        days: formatDays(group.days),
        hours: formatHours(group.hours)
      })
    ))

  // group days with same hours together
  const groupOpeningHours = (acc: OpeningHourGrouped[], curr: OpeningHour): OpeningHourGrouped[] => {
    const previous = acc.at(-1)
    // only group days with same opening hours together if they follow each other
    const hoursSameAsPervious = JSON.stringify(previous?.hours) === JSON.stringify(curr.hours)

    if (hoursSameAsPervious) {
      // add the current day to previous group if opening hours are the same
      previous?.days.push(curr.day)
    } else {
      // create new group if opening hours are not the same
      acc.push({
        id: curr.id,
        days: [curr.day],
        hours: curr.hours
      })
    }

    return acc
  }

  // format days to display the first and last day if multiple days are selected
  const formatDays = (days: Day[]): string => {
    const firstDay = days.at(0)
    const lastDay = days.at(-1)
    const hasMultipleDays = days.length > 1
    const daysFormatted = hasMultipleDays
      ? `${firstDay} - ${lastDay}`
      : firstDay

    return daysFormatted || ''
  }

  // format hours to show the opening hours if defined or the string 'closed' if not defined
  const formatHours = (hours: Hour[]): string[] => hours.length
    ? hours.map(time => `${time.start} - ${time.end}`)
    : ['closed']
</script>

<style lang="scss" scoped>
  .opening-hours {
    &__values {
      text-transform: capitalize;
    }
  }
</style>
