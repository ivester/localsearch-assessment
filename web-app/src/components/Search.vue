<template>
  <v-container class="fill-height">
    <v-responsive class="align-center text-center fill-height">
      <!-- TODO make into components -->
      <v-text-field
        label="Search Business Entries..."
        variant="solo-filled"
        append-inner-icon="mdi-magnify"
        @click:append-inner="search"
        @keyup.enter="search"
      />
      <v-card v-if="data.displayed_what">
        <v-card-title>
          <span class="headline">{{ data.displayed_what }}</span>
        </v-card-title>
      </v-card>
    </v-responsive>
  </v-container>
</template>

<script lang="ts" setup>
  import axios from 'axios'
  import { ref } from 'vue'

  interface Data {
    displayed_what: string
  }

  const data = ref<Data>({ displayed_what: '' })
  async function search() {
    const response = await axios.get<Data>('http://localhost:4000')
    data.value = response.data
  }
</script>
