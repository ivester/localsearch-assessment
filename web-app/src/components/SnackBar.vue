<template>
  <v-snackbar
    v-model="showSnackbar"
    class="snack-bar"
    location="top"
    color="red"
    :timeout="5000"
  >
    <div class="snack-bar__error">
      <v-icon
        class="snack-bar__error-icon"
        icon="mdi-alert"
      />
      <span>{{ message }}</span>
    </div>
  </v-snackbar>
</template>

<script lang="ts" setup>
import { ref, watch } from 'vue';

  interface Props {
    message: string;
    show: boolean;
  }

  const props = withDefaults(defineProps<Props>(), {
    show: false
  })

  const showSnackbar = ref<boolean>(props.show)

  watch(() => props.show, (value) => {
    showSnackbar.value = value
  })
</script>

<style scoped lang="scss">
  .snack-bar {
    &__error {
      display: flex;
      flex-direction: row;
      align-items: center;
      justify-content: center;

      &-icon {
        margin-right: 16px;
      }
    }
  }
</style>
