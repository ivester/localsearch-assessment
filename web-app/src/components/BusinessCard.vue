<template>
  <v-card
    :to="link"
    variant="tonal"
    :elevation="4"
  >
    <v-card-title>
      <span class="headline">{{ business.name }}</span>
    </v-card-title>
    <v-card-text>
      {{ business.where }} <br />
      <div v-if="mode === 'detail'">
        <span v-if="business?.url">Website: <a :href="business.url" target="_blank">{{ business?.urlFormatted || business.url }}</a> <br /></span>
        <span v-if="business?.phone">Phone: <a :href="`tel:${business.phone}`">{{ business?.phoneFormatted || business.phone }}</a> <br /></span>
        <OpeningHours v-if="business?.openingHours" :opening-hours="business.openingHours"/>
      </div>
    </v-card-text>
  </v-card>
</template>

<script lang="ts" setup>
  import { Business } from '@/types';
  import OpeningHours from './OpeningHours.vue';
import { computed } from 'vue';
import { RouteLocationRaw } from 'vue-router';

  // --- Props ---

  interface Props {
    business: Business;
    mode?: 'search' | 'detail';
  }

  const props = withDefaults(defineProps<Props>(), {
    mode: 'detail'
  })

  // --- Computed ---

  const link = computed((): RouteLocationRaw | undefined => {
    return props.mode === 'search' ? { name: 'Detail', params: { id: props.business.id } } : undefined
  })
</script>
