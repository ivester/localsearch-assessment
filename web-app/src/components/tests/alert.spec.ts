import { mount } from '@vue/test-utils'
import { expect, test } from 'vitest'
import { createVuetify } from 'vuetify'
import * as components from 'vuetify/components'
import * as directives from 'vuetify/directives'
import { createRouter, createWebHistory } from 'vue-router'
import { routes } from '@/router'
import Alert from '../Alert.vue'

const vuetify = createVuetify({
  components,
  directives,
})

const router = createRouter({
  history: createWebHistory(),
  routes,
})

global.ResizeObserver = require('resize-observer-polyfill')

test('shows alert', () => {
  const wrapper = mount(Alert, {
    props: { message: 'error message'},
    global: {plugins: [vuetify, router]}
  })

  expect(wrapper.html()).toMatchSnapshot()
})
