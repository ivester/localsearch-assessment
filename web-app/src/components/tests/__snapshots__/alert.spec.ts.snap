// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`shows alert 1`] = `
"<div data-v-3f4a8ec2=\\"\\" class=\\"v-alert v-theme--light bg-error v-alert--density-default v-alert--variant-flat\\" role=\\"alert\\">
  <!----><span class=\\"v-alert__underlay\\"></span>
  <!---->
  <div class=\\"v-alert__prepend\\"><i class=\\"mdi-close-circle mdi v-icon notranslate v-theme--light\\" style=\\"font-size: 28px; height: 28px; width: 28px;\\" aria-hidden=\\"true\\" density=\\"default\\"></i></div>
  <div class=\\"v-alert__content\\">
    <!---->error message
    <!-- div wrapper to show the button under the text and not inline with the text -->
    <div data-v-3f4a8ec2=\\"\\"><a data-v-3f4a8ec2=\\"\\" class=\\"v-btn v-btn--elevated v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-elevated alert-button\\" href=\\"/\\"><span class=\\"v-btn__overlay\\"></span><span class=\\"v-btn__underlay\\"></span>
        <!----><span class=\\"v-btn__content\\" data-no-activator=\\"\\">Go to Search</span>
        <!---->
        <!---->
      </a></div>
  </div>
  <!---->
  <!---->
</div>"
`;
